### PR TITLE
Add Pancakeswap NFT Events

### PR DESCRIPTION
**Pull request by thb030100**
Brief comments on the purpose of your changes:

**For Dune Engine V2**

I've checked that:

### General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test

### CODE REVIEW

File: models/pancakeswap/bnb/pancakeswap_bnb_nft_market_v1_event.sql
Comment (0xRobin) [inline models/pancakeswap/bnb/pancakeswap_bnb_nft_market_v1_event.sql:85]: This can be made an inner join for better performance
```suggestion
    INNER JOIN {{ source('bnb','transactions') }} bt ON bt.hash=events.tx_hash
```

File: models/pancakeswap/bnb/pancakeswap_bnb_nft_market_v1_event.sql
Comment (0xRobin) [inline models/pancakeswap/bnb/pancakeswap_bnb_nft_market_v1_event.sql:26]:
```suggestion
        '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c' AS currency_contract,
        'BNB' AS currency_symbol,
```

File: models/pancakeswap/bnb/pancakeswap_bnb_schema.yml
Comment (0xRobin) [inline models/pancakeswap/bnb/pancakeswap_bnb_schema.yml:106]: We should also add tests here:
```yml
  tests:
      - dbt_utils.unique_combination_of_columns:
          combination_of_columns:
            - block_date
            - unique_trade_id
      - check_seed:
          seed_file: ref('pancakeswap_nft_trades_samples')
          match_columns:
            - block_time
            - tx_hash
          check_columns:
            - token_id
            - amount_original
            - buyer
            - seller
```

File: models/pancakeswap/bnb/pancakeswap_bnb_nft_market_v1_event.sql
Comment (0xRobin) [inline models/pancakeswap/bnb/pancakeswap_bnb_nft_market_v1_event.sql:74]: we're trying to minimize the amount of columns used here.
`block_number - tx_hash - evt_index` will do in this case.
```suggestion
        events.block_number || events.tx_hash || events.evt_index AS unique_trade_id
```
You'll have to add the evt_index in the events CTE above.

**Review by 0xRobin — CHANGES_REQUESTED** [review]
Thanks for the PR, I've added some minor suggestions.
Main thing is we'd like to see a seed file with a couple example trades in them to setup a test.
You can add that under `seeds/pancakeswap/bnb/pancakeswap_nft_trades_samples.csv`.
Then you'll also have to update the `dbt_project.yml` file to include the seed.

Let me know if you have questions.

**Comment by thb030100** [comment]
> Thanks for the PR, I've added some minor suggestions. Main thing is we'd like to see a seed file with a couple example trades in them to setup a test. You can add that under `seeds/pancakeswap/bnb/pancakeswap_nft_trades_samples.csv`. Then you'll also have to update the `dbt_project.yml` file to include the seed.
> 
> Let me know if you have questions.

I have added the seed file, can you check! Thanks

**Comment by 0xRobin** [comment]
Looks like the test samples are not passing, @thb030100 can you check?

**Comment by thb030100** [comment]
> 

I don't know why the test fail. Can you check for me?

**Comment by 0xRobin** [comment]
Yes I'll have a look

**Comment by thb030100** [comment]
> Yes I'll have a look

Have you founded it?

**Comment by thb030100** [comment]
I temporary comment on the test section since I don't understand clearly how the test works
`with matched_records as (
        select
            seed.block_time as seed_block_time,
            model_sample.block_time as model_block_time,
            
            seed.tx_hash as seed_tx_hash,
            model_sample.tx_hash as model_tx_hash,
            
            seed.token_id as seed_token_id,
            model_sample.token_id as model_token_id ,
            
            seed.amount_original as seed_amount_original,
            model_sample.amount_original as model_amount_original ,
            
            seed.buyer as seed_buyer,
            model_sample.buyer as model_buyer ,
            
            seed.seller as seed_seller,
            model_sample.seller as model_seller 
            from test_data.pancakeswap_nft_trades_samples seed
        left join (
            select
                model.block_time,
                
                model.tx_hash,
                
                model.token_id ,
                
                model.amount_original ,
                
                model.buyer ,
                
                model.seller 
                from  test_data.pancakeswap_nft_trades_samples seed
            inner join pancakeswap_bnb.events model
                ON 1=1
                    AND seed.block_time = model.block_time
                    
                    AND seed.tx_hash = model.tx_hash
                    ) model_sample
        ON 1=1
            AND seed.block_time = model_sample.block_time
            
            AND seed.tx_hash = model_sample.tx_hash
            WHERE 1=1),
`
Following this code, I guest it compare with pancakeswap_bnb.events which not exists yet. This might be the issue

**Comment by 0xRobin** [comment]
@thb030100 hey I just pushed to re-enable the tests 
(also updated the columns of the model in the schema file)

I'll walk you through how we can evaluate what is happening. 
So the dbt slim ci process will produce temporary tables for both the model and the seed file.
These you can query on Dune. :astonished: 

So if a test is not passing or working as expected:
1. [local] dbt compile and track down the compiled test file
2. check the logs from the dbt slim ci to get the names of the temporary model and seed tables
3. replace the sources in the compiled test with the temporary tables.
4. Run on Dune and investigate the results.

I'll wait for the dbt slim ci to run and give more instructions, we can walk through this together.

**Comment by 0xRobin** [comment]
CC: @cryptofreedman we should probably describe this process in https://dune.com/docs/spellbook/getting-started/tests/ to help others.

**Comment by thb030100** [comment]
@0xRobin thank you so much for the instruction! I have changed the test file and let see how it goes. 
This pr requests was so tough lol

**Comment by 0xRobin** [comment]
I believe it's not picking the dates in the seed file correctly. The format should be 'YYYY-MM-DD hh:mm:ss'

**Comment by thb030100** [comment]
> I believe it's not picking the dates in the seed file correctly. The format should be 'YYYY-MM-DD hh:mm:ss'

Hey I have changed the format, it's YYYY-MM-DD hh:mm:ss now

2021-09-30 10:06:35

**Comment by 0xRobin** [comment]
@thb030100 looks like tests are passing! :clap: 
I will add some changes for consistency and integrate this model with the general nft.trades one.
After that we are good to merge this!
Thanks for the contribution, the first one is always the hardest. :smile:

**Comment by thb030100** [comment]
> @thb030100 looks like tests are passing! 👏
> I will add some changes for consistency and integrate this model with the general nft.trades one.
> After that we are good to merge this!
> Thanks for the contribution, the first one is always the hardest. 😄

Thank you! I am also adding a few more marketplaces. Will follow this PR as an instruction!

**Comment by dot2dotseurat** [comment]
Sorry, switching this to do not merge for the time being. I'm checking if this will work on DuneSQL and it will cause nft.trades to be unqueryable. I need to look more closely to see why. 

@thb030100 thank you for making this contribution, we'll get it merged and welcome more contributions. Excellent first contribution by the way!

**Comment by belen-pruvost** [comment]
Hey @dot2dotseurat! Yes, the underlying issue here is that `test_schema.git_653e5544_pancakeswap_bnb_nft_events` has a DBFS location instead of an S3 one (see [here](https://dbc-7ee55e77-eabe.cloud.databricks.com/sql/explore/data/hive_metastore/test_schema/git_653e5544_pancakeswap_bnb_nft_events?o=4190327660257548)). Ideally the CI that creates the test schema would need to use something like what we discussed [here](https://dune.height.app/T-22083).

**Comment by dot2dotseurat** [comment]
@belen-pruvost ah, makes sense. thanks for checking. 

We can probably go ahead and merge this then.

**Review by soispoke — APPROVED** [review]
lgtm

**Pull request merged**
